redis/src/lib.rs

```diff
@@ -0,0 +1,100 @@
+//! Redis support for the `bb8` connection pool.
+#![deny(missing_docs, missing_debug_implementations)]
+
+pub extern crate bb8;
+pub extern crate redis;
+
+extern crate futures;
+extern crate tokio_core;
+
+use futures::{Future, IntoFuture};
+use tokio_core::reactor::Handle;
+
+use redis::async::Connection;
+use redis::{Client, RedisError};
+
+use std::io;
+use std::option::Option;
+
+type Result<T> = std::result::Result<T, RedisError>;
+
+/// `RedisPool` is a convenience wrapper around `bb8::Pool` that hides the fact that
+/// `RedisConnectionManager` uses an `Option<Connection>` to smooth over the API incompatibility.
+#[derive(Debug)]
+pub struct RedisPool {
+    // The `bb8::Builder` that manages the pool.
+    pool: bb8::Pool<RedisConnectionManager>,
+}
+
+impl RedisPool {
+    /// Constructs a new `RedisPool`, see the `bb8::Builder` documentation for description of
+    /// parameters.
+    pub fn new(pool: bb8::Pool<RedisConnectionManager>) -> RedisPool {
+        RedisPool { pool }
+    }
+
+    /// Run the function with a connection provided by the pool.
+    pub fn run<'a, T, E, U, F>(&self, f: F) -> Box<Future<Item = T, Error = E> + Send + 'a>
+    where
+        F: FnOnce(Connection) -> U + Send + 'a,
+        U: IntoFuture<Item = (Connection, T), Error = E> + 'a,
+        U::Future: Send + 'a,
+        E: From<RedisError> + Send + 'a,
+        T: Send + 'a,
+    {
+        let f = move |conn: Option<Connection>| {
+            let conn = conn.unwrap();
+            f(conn)
+                .into_future()
+                .map(|(conn, item)| (item, Some(conn)))
+                .map_err(|err| (err, None))
```

Comment:
I don't like that on error the underlying connection is thrown away here. This precludes ever usefully implementing an equivalent to the `CustomizeConnection::on_release` functionality in r2d2. Is there a reason not to return the connection in the error case?

Reply:
Unfortunately, the underlying library swallows the connection on error:

https://docs.rs/redis/0.9.1/redis/struct.Cmd.html#method.query_async
https://docs.rs/redis/0.9.1/redis/type.RedisFuture.html

Reply:
Hrm ok :/